src/Servers/Kestrel/Core/src/Adapter/Internal/RawStream.cs

```diff
@@ -105,14 +107,19 @@ public override Task FlushAsync(CancellationToken cancellationToken)
             return WriteAsync(null, 0, 0, cancellationToken);
         }
 
-        private async ValueTask<int> ReadAsyncInternal(Memory<byte> destination)
+        private async ValueTask<int> ReadAsyncInternal(Memory<byte> destination, CancellationToken cancellationToken)
         {
             while (true)
             {
-                var result = await _input.ReadAsync();
+                var result = await _input.ReadAsync(cancellationToken);
                 var readableBuffer = result.Buffer;
                 try
                 {
+                    if (_throwOnCancelled && result.IsCanceled)
+                    {
+                        throw new OperationCanceledException();
```

Comment:
We need a flag to know *we're* the once that canceled the read.

Reply:
Done